In [1]:
#Impports necessários
import pandas as pd
import numpy as np

import mlflow
import mlflow.sklearn

from sklearn.model_selection import train_test_split

from sklearn.linear_model import (LinearRegression, Lasso)

from sklearn.metrics import (mean_absolute_error, mean_squared_error, r2_score)

In [2]:
#Configurações do MLflow

mlflow.set_tracking_uri("sqlite:///mlflow.db")

mlflow.set_experiment("Lasso_Regression")

mlflow.sklearn.autolog()

In [3]:
#Carregar dataset

df = pd.read_csv("../DATASET/dataset_expandido.csv")

In [4]:
#Preparação do Dataset - escolha das features e do targer

df["YearsCodePro_Num"] = (df["YearsCodePro"])

df["YearsCodePro_Num"] = (df["YearsCodePro_Num"].replace({"Less than 1 year":0, "More than 50 years":51, "Sem dado":0}))

df["YearsCodePro_Num"] = pd.to_numeric(df["YearsCodePro_Num"])

features = [
    "YearsCodePro_Num",
    "WorkExp",
    "Age_Code"
]

target = ("ConvertedCompYearly")

X = df[features]

y = df[target]

In [5]:
#Divisão do dataset

X_train, X_test, y_train, y_test = (train_test_split(X, y, test_size=0.2, random_state=42))

In [6]:
#Função do Lasso Regression

def run_lasso_experiment(
    run_name,
    alpha
):

    with mlflow.start_run(run_name=run_name):

        model = Lasso(alpha=alpha)

        model.fit(X_train, y_train)

        y_pred = model.predict(X_test)

        rmse = np.sqrt(mean_squared_error(y_test, y_pred))

        r2 = r2_score(y_test, y_pred)

        mlflow.log_param("alpha", alpha)

        mlflow.log_metric("RMSE", rmse)

        mlflow.log_metric("R2", r2)

        print(run_name)

        print("Alpha:", alpha)

        print("RMSE:", rmse)

        print("R2:",r2)

        print(model.coef_)
 
        print("-"*40)

        return model

In [25]:
# Experienncia 0 - Regressão Linear Baseline

linear_model = LinearRegression()

linear_model.fit(
    X_train,
    y_train
)

print(
    linear_model.coef_
)

2026/05/22 16:18:51 INFO mlflow.utils.autologging_utils: Created MLflow autologging run with ID 'f0951f6183674daa927965a6e82ff874', which will track hyperparameters, performance metrics, model artifacts, and lineage information for the current sklearn workflow
2026/05/22 16:18:52 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Local\Programs\Python\Python313\Lib\site-packages\mlflow\types\utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Inte

[799.34344869 804.43466193 162.22745526]


In [26]:
# Experiencia 1 - Lasso Regression com alpha = 0.01

lasso001 = run_lasso_experiment(
    run_name=
    "Experiment_1_Lasso_alpha_0.01",

    alpha=0.01
)

2026/05/22 16:18:59 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Local\Programs\Python\Python313\Lib\site-packages\mlflow\types\utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/05/22 16:18:59 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Lo

Experiment_1_Lasso_alpha_0.01
Alpha: 0.01
RMSE: 137888.47967394858
R2: 0.007214354632622211
[799.34439589 804.43468422 162.21584113]
----------------------------------------


In [27]:
# Experiencia 2 - Lasso Regression com alpha = 0.1

lasso01 = run_lasso_experiment(
    run_name=
    "Experiment_3_Lasso_alpha_0.1",

    alpha=0.1
)

2026/05/22 16:19:06 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Local\Programs\Python\Python313\Lib\site-packages\mlflow\types\utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/05/22 16:19:06 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Lo

Experiment_3_Lasso_alpha_0.1
Alpha: 0.1
RMSE: 137888.47980888103
R2: 0.007214352689617232
[799.35292074 804.43490162 162.1112718 ]
----------------------------------------


In [28]:
# Experiencia 3 - Lasso Regression com alpha = 1.0

lasso1 = run_lasso_experiment(
    run_name=
    "Experiment_4_Lasso_alpha_1",

    alpha=1
)

2026/05/22 16:19:14 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Local\Programs\Python\Python313\Lib\site-packages\mlflow\types\utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/05/22 16:19:14 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Lo

Experiment_4_Lasso_alpha_1
Alpha: 1
RMSE: 137888.4811640554
R2: 0.0072143331753302675
[799.43807133 804.4371263  161.06595779]
----------------------------------------


In [29]:
# Experiencia 4 - Lasso Regression com alpha = 10.0

lasso10 = run_lasso_experiment(
    run_name=
    "Experiment_5_Lasso_alpha_10",

    alpha=10
)

2026/05/22 16:19:24 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Local\Programs\Python\Python313\Lib\site-packages\mlflow\types\utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/05/22 16:19:24 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Lo

Experiment_5_Lasso_alpha_10
Alpha: 10
RMSE: 137888.49504500197
R2: 0.007214133291984726
[800.29021689 804.46063868 150.60637718]
----------------------------------------


In [30]:
# Experiencia 5 - Lasso Regression com alpha = 100

lasso100 = run_lasso_experiment(
    run_name=
    "Experiment_6_Lasso_alpha_100",

    alpha=100
)

2026/05/22 16:19:32 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Local\Programs\Python\Python313\Lib\site-packages\mlflow\types\utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/05/22 16:19:32 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Lo

Experiment_6_Lasso_alpha_100
Alpha: 100
RMSE: 137888.6679745339
R2: 0.007211643133500645
[808.83629542 804.69467557  45.88630359]
----------------------------------------


In [31]:
# Experiencia 6 - Lasso Regression com alpha = 1000

lasso1000 = run_lasso_experiment(
    run_name=
    "Experiment_7_Lasso_alpha_1000",

    alpha=1000
)

2026/05/22 16:19:45 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Local\Programs\Python\Python313\Lib\site-packages\mlflow\types\utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2026/05/22 16:19:45 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "c:\Users\luisr\AppData\Lo

Experiment_7_Lasso_alpha_1000
Alpha: 1000
RMSE: 137889.28404557146
R2: 0.0072027717805687175
[810.16231344 785.41641497   0.        ]
----------------------------------------


In [32]:
#Tabela de Comparação de resultados

comparison = pd.DataFrame({

    "Feature":
    features,

    "Linear":
    linear_model.coef_,

    "Lasso_0.01":
    lasso01.coef_,

    "Lasso_0.1":
    lasso01.coef_,

    "Lasso_1":
    lasso1.coef_,

    "Lasso_10":
    lasso10.coef_,

    "Lasso_100":
    lasso100.coef_,

    "Lasso_1000":
    lasso1000.coef_

})

comparison

,Feature,Linear,Lasso_0.01,Lasso_0.1,Lasso_1,Lasso_10,Lasso_100,Lasso_1000
0,YearsCodePro_Num,799.343449,799.352921,799.352921,799.438071,800.290217,808.836295,810.162313
1,WorkExp,804.434662,804.434902,804.434902,804.437126,804.460639,804.694676,785.416415
2,Age_Code,162.227455,162.111272,162.111272,161.065958,150.606377,45.886304,0.000000


### Introdução

A Lasso Regression foi aplicada utilizando as mesmas variáveis usadas anteriormete na Regressão Linear, mantendo a consistência entre os modelos avaliados.

Foram testados diferentes valores para o parâmetro alpha:

    -0.01
    -0.1
    -1
    -10
    -100
    -1000

O objetivo é analisar o impacto da regularização L1 e verificar a possibilidade de eliminação de variáveis. Ao contrário da Ridge Regression, observou-se uma degradação progressiva do desempenho à medida que alpha aumentou.

### ANÁLISE DE EXPERIÊNCIAS

# Experiência 1 - Regressão Linear Baseline

A regressão Linear Baseline serviu como modelo de referência. O modelo apresentou como resultados:

    .R^2 -> 0.007
    .RMSE -> 137888

Estes resultados ajudam a confirmar que o modelo tem uma capacidade preditiva muito baixa das variáveis selecionadas. 

# Experiência 2 - Ridge alpha = 0.01

Devido ao baixo valor de alpha, as diferenças entre os resultados foram quase nulas. Neste cenário a penalização L1 possui impacto reduzido, preservando praticamente todos os coeficientes.

# Experiência 3 - Ridge alpha = 0.1

A partir deste valor de alpha, começaram a observar-se maiores alterações dos resultados. Verificamos uma ligeira redução do desempenho. Este comportamento mostra-nos que o Lasso começou a atuar sobre os coeficientes. 

# Experiência 4 - Ridge alpha = 1

Com o valor = 1, continuamos a verificar a mesma tendência que os anteriores. os resultados continuam  a mudar de forma lenta, indicando o aumento da penalização.

# Experiência 5 - Ridge alpha = 10

Começamos a implementar valores mais altos, e aqui observamos maior variância nos resultados. Verificamos que o RMSE aumentou e que o R^2 diminui ligeiramente, o que não são bons resultados. Ou seja, depois desta análise entendemos que o modelo começou a perderr capacidade explicativa.

# Experiência 6 - Ridge alpha = 100

Seguimos a sequencia, com os valores sempre a mudar. Os resultados mostram que o aumento da regularização passou a reduzir excessivamente a informação utilizada pelo modelo.

# Experiência 7 - Ridge alpha = 1000

Esta experiência foi a que apresentou os piores resultados, como era de esperar pela análise que temos vindo a fazer. Mais uma vez com um valor muito mais alto de alpha, os resultados pioraram. Apesar da degradação não ser extrema, observou-se perda gradual de desempenho.

Este comportamento era esperado, pois valores elevados de alpha provocam forte penalização dos coeficientes.

# Análise de Coeficientes

O principal objetivo do Lasso não é apenas melhorar métricas, mas também realizar uma seleção automática de variáveis. Esta característica torna o modelo particularmente útil em problemas com muitas variáveis, permitindo reduzir complexidade e facilitar interpretação.

No presente trabalho, como apenas foram utilizadas três variáveis explicativas, o efeito de eliminação tornou-se naturalmente mais limitado. Mesmo assim, observou-se que o aumento de alpha produziu maior impacto do que no Ridge, demonstrando que a regularização L1 é mais agressiva.

### Conclusão

Os resultados mostram que a Regressão Lasso apresentou comportamento mais dependente do parâmetro alpha quando comparada ao Ridge. Os valores reduzidos mantiveram desempenho semelhante ao baseline, enquanto valores elevados provocaram perda gradual de capacidade preditiva.

Este comportamento confirma a teoria associada à regularização L1, que tenta reduzir a complexidade do modelo através da penalização e a  possível eliminação de variáveis. 

# Hyperparameter Optimization - Lasso 

In [8]:
import mlflow
import optuna

from sklearn.model_selection import (
    train_test_split,
    GridSearchCV,
    cross_val_score
)

from sklearn.linear_model import Lasso

In [9]:
target = "ConvertedCompYearly"

features = [

    "YearsCodePro_Num",

    "WorkExp",

    "Age_Code"

]

X = df[features]

y = df[target]

X_train, X_test, y_train, y_test = train_test_split(

    X,

    y,

    test_size=0.2,

    random_state=42

)

In [10]:
mlflow.sklearn.autolog(

    disable=True

)

In [11]:
param_grid = {

    "alpha":[

        0.001,

        0.01,

        0.1,

        1,

        10,

        100

    ]

}

grid = GridSearchCV(

    estimator=Lasso(

        max_iter=1000

    ),

    param_grid=param_grid,

    cv=10,

    scoring="r2",

    n_jobs=-1

)

grid.fit(

    X_train,

    y_train

)

print(

    "Best Parameters:",

    grid.best_params_

)

print(

    "Best Score:",

    grid.best_score_

)

Best Parameters: {'alpha': 0.001}
Best Score: 0.02475758315329647


In [12]:
with mlflow.start_run(

    run_name="GridSearch_Lasso"

):

    mlflow.log_param(

        "method",

        "GridSearch"

    )

    mlflow.log_params(

        grid.best_params_

    )

    mlflow.log_metric(

        "best_score",

        grid.best_score_

    )

2026/05/28 21:36:36 WARNING mlflow.utils.git_utils: Failed to import Git (the Git executable is probably not on your PATH), so Git SHA is not available. Error: Failed to initialize: Bad git executable.
The git executable must be specified in one of the following ways:
    - be included in your $PATH
    - be set via $GIT_PYTHON_GIT_EXECUTABLE
    - explicitly set via git.refresh(<full-path-to-git-executable>)

All git commands will error until this is rectified.

This initial message can be silenced or aggravated in the future by setting the
$GIT_PYTHON_REFRESH environment variable. Use one of the following values:
    - quiet|q|silence|s|silent|none|n|0: for no message or exception
    - warn|w|warning|log|l|1: for a warning message (logging level CRITICAL, displayed by default)
    - error|e|exception|raise|r|2: for a raised exception

Example:
    export GIT_PYTHON_REFRESH=quiet



In [13]:
def objective(

    trial

):

    alpha = trial.suggest_categorical(

        "alpha",

        [

            0.001,

            0.01,

            0.1,

            1,

            10,

            100

        ]

    )

    model = Lasso(

        alpha=alpha,

        max_iter=1000

    )

    scores = cross_val_score(

        model,

        X_train,

        y_train,

        cv=10,

        scoring="r2",

        n_jobs=-1

    )

    return scores.mean()

In [14]:
study = optuna.create_study(

    direction="maximize"

)

study.optimize(

    objective,

    n_trials=100

)

print(

    "Best Parameters:",

    study.best_params

)

print(

    "Best Score:",

    study.best_value

)

[I 2026-05-28 21:37:09,588] A new study created in memory with name: no-name-6113e9b9-0595-40b7-b479-ca03b6f7cf9f
[I 2026-05-28 21:37:10,027] Trial 0 finished with value: 0.024757579889876525 and parameters: {'alpha': 0.1}. Best is trial 0 with value: 0.024757579889876525.
[I 2026-05-28 21:37:10,447] Trial 1 finished with value: 0.024757582856249415 and parameters: {'alpha': 0.01}. Best is trial 1 with value: 0.024757582856249415.
[I 2026-05-28 21:37:10,842] Trial 2 finished with value: 0.02475723506812183 and parameters: {'alpha': 10}. Best is trial 1 with value: 0.024757582856249415.
[I 2026-05-28 21:37:11,277] Trial 3 finished with value: 0.02475758315329647 and parameters: {'alpha': 0.001}. Best is trial 3 with value: 0.02475758315329647.
[I 2026-05-28 21:37:11,683] Trial 4 finished with value: 0.024757582856249415 and parameters: {'alpha': 0.01}. Best is trial 3 with value: 0.02475758315329647.
[I 2026-05-28 21:37:12,100] Trial 5 finished with value: 0.02475758315329647 and parame

Best Parameters: {'alpha': 0.001}
Best Score: 0.02475758315329647


In [15]:
with mlflow.start_run(

    run_name="Optuna_Lasso"

):

    mlflow.log_param(

        "method",

        "Optuna"

    )

    mlflow.log_params(

        study.best_params

    )

    mlflow.log_metric(

        "best_score",

        study.best_value

    )